File: src/executors/graphql.ts
import type { Tool, ToolExecutor, ToolExecutionResult } from '../types.js';

function buildQuery(tool: Tool, args: Record<string, unknown>): string {
  const kind = tool.method; // 'query' or 'mutation'
  const params = tool.parameters;

  if (params.length === 0) {
    return `${kind} { ${tool.name} }`;
  }

  // Build variable definitions and arguments
  const varDefs = params.map(p => `$${p.name}: ${p.type}${p.required ? '!' : ''}`).join(', ');
  const fieldArgs = params
    .filter(p => args[p.name] !== undefined)
    .map(p => `${p.name}: $${p.name}`)
    .join(', ');

  return `${kind}(${varDefs}) { ${tool.name}(${fieldArgs}) }`;
}

export const graphqlExecutor: ToolExecutor = {
  protocol: 'graphql',

  async execute(tool: Tool, args: Record<string, unknown>): Promise<ToolExecutionResult> {
    const query = buildQuery(tool, args);
    const variables: Record<string, unknown> = {};
    for (const param of tool.parameters) {
      if (args[param.name] !== undefined) {
        variables[param.name] = args[param.name];
      }
    }

    const res = await fetch(tool.endpoint, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ query, variables }),
    });

    const raw = await res.text();
    let data: unknown;
    try { data = JSON.parse(raw); } catch { data = raw; }

    const headers: Record<string, string> = {};
    res.headers.forEach((value, key) => { headers[key] = value; });

    return { status: res.status, data, headers, raw };
  },
};

File: src/api_to_tools/executors/graphql.py
"""GraphQL executor."""

from __future__ import annotations

import httpx

from api_to_tools.constants import DEFAULT_EXECUTOR_TIMEOUT
from api_to_tools.types import AuthConfig, Tool, ExecutionResult


def _build_query(tool: Tool, args: dict) -> str:
    kind = tool.method  # 'query' or 'mutation'
    selection_set = tool.metadata.get("selection_set", "")

    used_params = [p for p in tool.parameters if p.name in args]
    if not used_params:
        return f"{kind} {{ {tool.name} {selection_set} }}"

    var_defs = ", ".join(f"${p.name}: {p.type}{'!' if p.required else ''}" for p in used_params)
    field_args = ", ".join(f"{p.name}: ${p.name}" for p in used_params)

    return f"{kind}({var_defs}) {{ {tool.name}({field_args}) {selection_set} }}"


def execute_graphql(tool: Tool, args: dict, *, auth: AuthConfig | None = None) -> ExecutionResult:
    """Execute a GraphQL query/mutation."""
    from api_to_tools.auth import build_auth_headers, build_auth_cookies, resolve_auth

    query = _build_query(tool, args)
    variables = {p.name: args[p.name] for p in tool.parameters if p.name in args}

    headers = {"Content-Type": "application/json"}
    cookies = {}
    if auth:
        resolved = resolve_auth(auth)
        headers.update(build_auth_headers(resolved))
        cookies = build_auth_cookies(resolved)

    verify = auth.verify_ssl if auth else True
    response = httpx.post(
        tool.endpoint,
        json={"query": query, "variables": variables},
        headers=headers,
        cookies=cookies or None,
        timeout=DEFAULT_EXECUTOR_TIMEOUT,
        verify=verify,
    )

    data = response.json()
    return ExecutionResult(
        status=response.status_code,
        data=data,
        headers=dict(response.headers),
        raw=response.text,
    )
